### DISCUSSION TITLE
当我用qt5.15.2 播放pag 动画崩溃d3d 中？升级的显卡驱动也无法解决

### DISCUSSION THREAD

**Discussion by weijibin-tal** [Tencent/libpag#2626] [Q&A] [answered]
![Uploading Snipaste_2024-12-17_20-51-57.png…]()
<img width="717" alt="Snipaste_2024-12-17_20-50-27" src="https://github.com/user-attachments/assets/1310fbf1-7fee-4000-b465-bf25e11bf366" />

**Comment by weijibin-tal** [comment]
![Uploading Snipaste_2024-12-17_20-51-57.png…]()

**Reply by domchen** [reply]
你应该是没有把QT的DLL目录加入Path，仔细看一下README文档

**Reply by weijibin-tal** [reply]
<img width="787" alt="Snipaste_2024-12-18_10-28-12" src="https://github.com/user-attachments/assets/23dc3111-bf7f-4a8a-8b08-07c04af182d5" /> 这个是windbg 的堆栈，我这边使用的安装程序安装的，这个问题只有部分电脑有问题，大部分电脑是正常的

**Reply by domchen** [reply]
按照这个步骤设置，确保都设置了：https://github.com/Tencent/libpag/blob/main/qt/README.zh_CN.md

**Reply by weijibin-tal** [reply]
我这是一个独立的安装程序，发布到其他电脑上使用的，，客户电脑上还要安装qt ?

**Reply by domchen** [reply]
你现在是自己开发了一个QT程序使用了PAG的库是麽？调试PAG就是上面的文档。发布安装程序给别人使用这个不是PAG范畴的技术问题，请去QT官网看文档提问，是要打包QT的DLL进去的。

**Reply by weijibin-tal** [reply]
我们自己编译的pag 的库 再我们程序 中使用，然后打包 发布，在部分电脑出现了这个问题，我们用的 pag::PAGSurface::MakeOffscreen 的得到，pag 每一帧的数据，或者有其他方法吗在qt（除了使用QQuick） 中得到 pag 中的每一帧的数据

**Reply by weijibin-tal** [reply]
> 你现在是自己开发了一个QT程序使用了PAG的库是麽？调试PAG就是上面的文档。发布安装程序给别人使用这个不是PAG范畴的技术问题，请去QT官网看文档提问，是要打包QT的DLL进去的。

是的 已经打包进去了，只有部分电脑出现这个问题， 从堆栈看是崩溃到 pag::PAGSurface::MakeOffscreen 这个函数中了

**Reply by domchen** [reply]
你这个是QT开发的问题，提问错了地方。PAG只是适配了QT库，你现在的问题是要打包发布一个QT程序，运行过程找不到DLL，只要调试通过这个就和PAG的代码没什么关系。打包的问题这个要去QT那边看打包发布相关的文档。Android 和 iOS 开发者也是对应去各自的官网看文档，PAG这里不会教你怎么开发自己的应用程序以及怎么打包发布App。

**Reply by weijibin-tal** [reply]
你怎么判断出 是缺少dll 的？ 你的意思是因为少库导致 初始化opengl环境报错的吗？ 现在是pag 内部初始化opengl 崩溃的，pag 里面不是用的tgfx 的渲染引擎吗？， 没有用到qt 的渲染引擎啊

**Reply by domchen** [reply]
你报错的第一行就说了缺少DLL啊：igd10um32xe.DLL！［下面的框架可能不正确和/或缺失，没有为 igd10um32xe.DLL加载符号］。你用了PAG或者tgfx，它渲染要OpenGL的环境，这个OpenGL的环境是QT提供的。但是我再仔细看了一下你的报错，是挂在了EGL的后端上，你现在是在用QT开发，为什么会用到EGL的后端？应该是编译错了库的版本。QT的版本要参考我刚刚发的那个文档。

**Reply by weijibin-tal** [reply]
我这个程序 启动都是正常的，只是运行过程中 一播放pag 动画就崩溃，怎么会缺少dll

**Reply by domchen** [reply]
另外下次PAG问题不要在别的项目提问，在PAG仓库那边提问就行。

**Reply by domchen** [reply, marked as answer]
> 我这个程序 启动都是正常的，只是运行过程中 一播放pag 动画就崩溃，怎么会缺少dll

缺的不是QT的DLL，是EGL的，运行到EGL初始化的时候会去访问EGL的动态库获取指针，在这个步骤找不到对应的DLL挂掉了。你是编译错了库的版本，似乎把Linux或者Windows平台的PAG库嵌入到QT程序里使用了。要编译PAG的QT版本的库，参考PAG的qt目录的demo，先跑通那个。主程序是哪个UI框架就需要用那个UI框架提供的OpenGL环境进行渲染，你编译了一个不是QT版本的PAG库在使用，也渲染不到你的QT界面上。

**Reply by domchen** [reply]
@weijibin-tal 你的程序如果目标并不是渲染内容到QT界面上，只是想读取离屏像素数据。也可以用SWIFTSHADERd的后端，这个是不依赖硬件GPU的，靠CPU模拟的OpenGL环境。编译库的时候传入-DPAG_USE_SWIFTSHADER=ON就行。另外一个解法，检查一下报错的那个dll具体是什么功能，你现在应该是编译了Windows平台原生的PAG库版本，那个是用到了ANGLE这个后端模拟的。你首先需要把ANGLE对应的那些EGL动态库都打包到安装程序，但是不确定ANGLE是不是还依赖的其他的系统DLL，所以要去检查一下报错的那个DLL具体是什么系统版本提供的，以及能不能也打包进去安装程序。最后一个解法就是正常使用QT的渲染后端，编译PAG for QT的版本进行渲染。

**Reply by weijibin-tal** [reply]
好的 ，我试试

**Reply by domchen** [reply]
不管用SwiftShader后端还是ANGLE后端，都要把对应的EGL相关的动态库打包到一起。你看看CMAKE配置里的引用路径。

**Reply by weijibin-tal** [reply]
编译了一个 qt 的版本 已经解决了，应该是渲染后端的不一样导致的崩溃

**Reply by domchen** [reply]
好的
